packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, OnPatchesDisposer, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposer: OnPatchesDisposer
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
+    if (isRootStore(subtreeRoot)) {
```

Comment:
Just made `isRootStore` reactive in 0.29.0, so it might be worth to do this like
```
let wasRS = false;
reaction(() => isRootStore(subTreeRoot), (isRS) => {
  if (isRS !== wasRS) {
    wasRS = isRS
    if (isRS) { registerRootStore(this.subtreeRoofDraft) }
    else { unregisterRootStore(this.subtreeRoofDraft) }
 }
  }
}, {fireImmediately: true})
```
That way the draft will follow the state of the original one

Reply:
That's a great idea! Done in 7b2cad4dfd9418dcd72347359ecc0b2ecfe1ea2a.

Do you like the way I'm dealing with the disposers? I was thinking about using an array of disposers (the way it's done now) or assigning a closure with the disposers to the private `disposer` class property. By using the array, I can remove the disposers from the array once they have been called - not sure if that's really necessary. In general, what behavior do you expect when a disposer is called multiple times?

Reply:
It should do nothing when called more than once, so that approach seems OK to me